types/yaireo__tagify/index.d.ts

```diff
@@ -947,13 +947,18 @@ declare namespace Tagify {
      */
     interface FocusEventData<T extends BaseTagData = TagData> extends FocusChangeEventData<T> { }
 
+    /**
+     * Possible error messages that will occur on an InvalidTagEvent
+     */
+    type InvalidErrorMessages = 'empty' | 'number of tags exceeded' | 'pattern mismatch' | 'already exists' | 'not allowed';
+
     /**
      * A tag has been added but did not pass validation.
      * @template T Type of the tag data. See the Tagify class for more details.
      */
     interface InvalidTagEventData<T extends BaseTagData = TagData> extends TagEventData<T> {
         data: T;
-        message: boolean;
+        message: InvalidErrorMessages | string | boolean;
```

Comment:
I certainly get the intention of specifying the expected messages while still allowing arbitrary strings, but type-wise, `InvalidErrorMessage | string` is the same as just `string`. Intellisense often treats it as just `string` anyway. I think what we should do here depends on how certain we are the `InvalidErrorMessages` are all there is. If (a) we can be certain there won't be any other messages, just change the type `InvalidErrorMessages | boolean`. Otherwise, the type should just `string | boolean` -- and the common error messages `empty`, `number of tags exceeded` etc. should be briefly mentioned in the comment for the `message` property.

```javascript
/**
 * Message indicating the type of error. Can be either a boolean to indicate success,
 * or a message code. Common message codes are:
 * - `empty` - When the new tag is empty.
 * - `number of tags exceeded` - When the new tag cannot be added because doing so would exceed the maximum number of allowed tags.
 * - `pattern mismatch` - When the new tag does not conform to the specified (regex) pattern.
 * - `already exists` - When a tag with the same name as the new tag exists already and duplicates are not allowed.
 * - `not allowed` -  When the new tag is not allowed for any other reason.
 */
message: string | boolean;
```

Reply:
My worry in making it only the specified errors it'll make newer error messages be impossible handle. But with `InvalidErrorMessages | string` it allows you to put anything however it also allows you to auto-input in most modern typescript supporting code editors. For example on vscode you can just put a '' and hit Ctrl+I and it'll show all of those strings there that'll let you auto-fill with them. Though you're correct that it won't make a different in compiling, just thought it'd be nice for the intellisense to suggest the errors that are atleast known, no?

Reply:
Yes, I understand it's for intellisense  / autocomplete only.  But in my experience, that does not really work as work that well, if at all. For example, in the current version of VSCode (and the typescript playground), you get this:

No autocomplete with a union, since it (correctly treats it as just `string`):
![image](https://user-images.githubusercontent.com/2456413/124428309-d5511780-dd6c-11eb-9787-96c2ae548a83.png)

Autocomplete works when the union contains only the messages:
![image](https://user-images.githubusercontent.com/2456413/124428415-f9acf400-dd6c-11eb-9780-d3cfa605c963.png)

Is you experience different? Users will usually consume the interface by reading the message, so I'd say specifying the messages as a comment is probably enough. And I'd tend to agree there's a good change more messages will be added, so 'string | boolean' sounds better.

Reply:
Interesting, I just tried it and you're correct. I would've sworn that it handled it like that. I'll fix it tomorrow.

Reply:
Thanks for your efforts! Please keep in mind that users can specify their custom validation functions returning custom error messages: [L649-L660](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/yaireo__tagify/index.d.ts#L649-L660). So I think `string | boolean` should be sufficient having the standard messages in the comment (as @blutorange suggested). Otherwise we would need to introduce another generic, so users can extend `InvalidErrorMessages`, which, however, would be over the top in my opinion. But that's arguable.

Reply:
Yeah, that's arguable, but in my experience, too much generics quickly become a burden and ends up being very hard to maintain (see some JQuery functions...) So `string | boolean` it is